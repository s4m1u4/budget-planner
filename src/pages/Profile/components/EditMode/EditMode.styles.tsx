import styled from "@emotion/styled";

export const ButtonGroup = styled.div({
  display: "flex",
  columnGap: "10px",
});
