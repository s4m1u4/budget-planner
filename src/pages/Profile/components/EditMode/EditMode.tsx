import React, { Component } from "react";
import { Formik, Form } from "formik";
import { EditModeSchema } from "./EditModeSchema";
import { IUserData } from "../../../../types";
import { ButtonComponent, InputComponent } from "../../../../components/shared";
import { PasswordModal } from "../PasswordModal";
import { IPasswordData } from "../../../Records/types";

import { ButtonGroup } from "./EditMode.styles";

interface IHandleSubmitValues {
  firstName: string;
  lastName: string;
  email: string;
}

interface EditModeProps {
  onSubmit: (values: IHandleSubmitValues) => void;
  editMode: string;
  navigate: (path: string | number) => void;
  userData: IUserData;
  setUserData: () => void;
  setNewUserData: (userData: IUserData) => void;
  setNewPassword: (passwordData: IPasswordData) => string | null;
}

interface EditModeState {
  isOpenModal: boolean;
}

export class EditMode extends Component<EditModeProps, EditModeState> {
  constructor(props: EditModeProps) {
    super(props);
    this.state = {
      isOpenModal: false,
    };
  }

  handleOpen = () => this.setState({ isOpenModal: true });
  handleClose = () => this.setState({ isOpenModal: false });

  handleClickEdit = () => {
    this.props.navigate("edit");
  };

  handleClickCancel = () => {
    this.props.navigate(-1);
  };

  handleSubmit = async (values: IHandleSubmitValues) => {
    const userData: IUserData = {
      ...this.props.userData,
      ...values,
    };
    await this.props.setNewUserData(userData);
    this.props.navigate(-1);
    await this.props.setUserData();
    this.props.onSubmit(values);
  };

  handleClickChangePassword = () => {
    this.handleOpen();
  };

  render() {
    const { firstName, lastName, email } = this.props.userData;

    return (
      <Formik
        enableReinitialize
        validationSchema={EditModeSchema}
        initialValues={{
          firstName: firstName,
          lastName: lastName,
          email: email,
        }}
        validateOnBlur
        onSubmit={async (values) => {
          await this.handleSubmit(values);
        }}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleBlur,
          handleSubmit,
          dirty,
        }) => (
          <Form style={{ width: "100%", maxWidth: "300px" }}>
            <PasswordModal
              open={this.state.isOpenModal}
              handleClose={this.handleClose}
              setNewPassword={this.props.setNewPassword}
              onSubmit={(values) => values}
            />
            <InputComponent
              fullWidth
              id="firstName"
              disabled={!Boolean(this.props.editMode)}
              label="First name"
              type="text"
              name="firstName"
              onBlur={handleBlur}
              onChange={handleChange}
              value={values.firstName || ""}
              error={(touched.firstName && !!errors.firstName) || false}
              helperText={(touched.firstName && errors.firstName) || ""}
            />
            <InputComponent
              fullWidth
              id="lastName"
              disabled={!Boolean(this.props.editMode)}
              label="Last name"
              type="text"
              name="lastName"
              onBlur={handleBlur}
              onChange={handleChange}
              value={values.lastName || ""}
              error={(touched.lastName && !!errors.lastName) || false}
              helperText={(touched.lastName && errors.lastName) || ""}
            />
            <InputComponent
              fullWidth
              id="email"
              disabled={!Boolean(this.props.editMode)}
              label="Email"
              type="text"
              name="email"
              onBlur={handleBlur}
              onChange={handleChange}
              value={values.email || ""}
              error={(touched.email && !!errors.email) || false}
              helperText={(touched.email && errors.email) || ""}
            />
            <ButtonComponent
              fullWidth
              style={{ marginBottom: "1rem" }}
              type="button"
              color="secondary"
              onClick={this.handleClickChangePassword}
            >
              Change password
            </ButtonComponent>
            <ButtonGroup>
              {!this.props.editMode ? (
                <ButtonComponent
                  fullWidth
                  type="button"
                  color="primary"
                  onClick={this.handleClickEdit}
                >
                  Edit
                </ButtonComponent>
              ) : (
                <>
                  <ButtonComponent
                    fullWidth
                    type="submit"
                    color="success"
                    disabled={!dirty}
                    onClick={handleSubmit}
                  >
                    Save
                  </ButtonComponent>
                  <ButtonComponent
                    fullWidth
                    type="button"
                    color="error"
                    disabled={dirty}
                    onClick={this.handleClickCancel}
                  >
                    Cancel
                  </ButtonComponent>
                </>
              )}
            </ButtonGroup>
          </Form>
        )}
      </Formik>
    );
  }
}
