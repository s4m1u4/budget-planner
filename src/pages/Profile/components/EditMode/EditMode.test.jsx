import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { EditMode } from "./EditMode";
import userEvent from "@testing-library/user-event";

const onSubmit = jest.fn().mockImplementationOnce((cb) => cb(null, true));
const navigate = jest.fn().mockImplementationOnce((cb) => cb(null, true));
const setUserData = jest.fn();
const setNewUserData = jest.fn();
const setNewPassword = jest.fn();
const editMode = "editMode";
const userData = {
  firstName: "Vadym",
  lastName: "Samitashvili",
  email: "totalgol2015@gmail.com",
};

const getInputFirstName = () => screen.getByLabelText("First name");
const getInputLastName = () => screen.getByLabelText("Last name");
const getInputEmail = () => screen.getByLabelText("Email");
const getButtonChangePassword = () =>
  screen.getByRole("button", { name: /Change password/i });
const getButtonEdit = () => screen.getByRole("button", { name: /Edit/i });
const getButtonSave = () => screen.getByRole("button", { name: /Save/i });
const getButtonCancel = () => screen.getByRole("button", { name: /Cancel/i });

const renderComponentWithEditMode = async () => {
  await act(
    async () =>
      await render(
        <EditMode
          onSubmit={onSubmit}
          editMode={editMode}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      )
  );
};

describe("Edit mode component", () => {
  describe("regular mode", () => {
    it("make snapshot", () => {
      const { baseElement } = render(
        <EditMode
          onSubmit={onSubmit}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );

      expect(baseElement).toMatchSnapshot();
    });

    beforeEach(() => {
      render(
        <EditMode
          onSubmit={onSubmit}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );
    });

    describe("should render", () => {
      it("an input with name='firstName'", () => {
        expect(getInputFirstName()).toHaveAttribute("name", "firstName");
      });

      it("an input with name='lastName'", () => {
        expect(getInputLastName()).toHaveAttribute("name", "lastName");
      });

      it("an input with name='email'", () => {
        expect(getInputEmail()).toHaveAttribute("name", "email");
      });

      it('a button "Change password"', () => {
        expect(getButtonChangePassword()).toBeInTheDocument();
      });

      it('a button "Edit"', () => {
        expect(getButtonEdit()).toBeInTheDocument();
      });
    });

    describe('modal "Create a new password"', () => {
      it("should open", () => {
        userEvent.click(getButtonChangePassword());
        expect(screen.getByText("Create a new password")).toBeInTheDocument();
      });

      it("should close", () => {
        userEvent.click(getButtonChangePassword());
        expect(screen.getByText("Create a new password")).toBeInTheDocument();
        userEvent.click(getButtonCancel());
        expect(screen.queryByText("Create a new password")).toBeNull();
      });
    });

    it("should go to page 'Edit mode'", () => {
      userEvent.click(getButtonEdit());
      render(
        <EditMode
          onSubmit={onSubmit}
          userData={userData}
          navigate={navigate}
          editMode={editMode}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );
      expect(screen.getByText(/cancel/i)).toBeInTheDocument();
    });
  });

  describe("edit mode", () => {
    it("make snapshot", () => {
      const { baseElement } = render(
        <EditMode
          onSubmit={onSubmit}
          editMode={editMode}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );

      expect(baseElement).toMatchSnapshot();
    });

    beforeEach(() => {
      render(
        <EditMode
          onSubmit={onSubmit}
          editMode={editMode}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );
    });

    describe("should render", () => {
      it('a button "Change password"', () => {
        expect(getButtonSave()).toBeInTheDocument();
      });

      it('a button "Cancel"', () => {
        expect(getButtonCancel()).toBeInTheDocument();
      });

      it("error message when submitting a form with empty inputs", async () => {
        expect(screen.queryByText("First name is required")).toBeNull();
        expect(screen.queryByText("Last name is required")).toBeNull();
        expect(screen.queryByText("Email is required")).toBeNull();
        fireEvent.change(getInputFirstName(), { target: { value: "" } });
        fireEvent.change(getInputLastName(), { target: { value: "" } });
        fireEvent.change(getInputEmail(), { target: { value: "" } });
        fireEvent.click(getButtonSave());
        await renderComponentWithEditMode();
        expect(screen.getByText("First name is required")).toBeInTheDocument();
        expect(screen.getByText("Last name is required")).toBeInTheDocument();
        expect(screen.getByText("Email is required")).toBeInTheDocument();
      });
    });

    it("should go to page 'Regular mode'", () => {
      userEvent.click(getButtonCancel());
      render(
        <EditMode
          onSubmit={onSubmit}
          userData={userData}
          navigate={navigate}
          setUserData={setUserData}
          setNewUserData={setNewUserData}
          setNewPassword={setNewPassword}
        />
      );
      expect(screen.getByText(/edit/i)).toBeInTheDocument();
    });

    it("successful form submission", async () => {
      fireEvent.change(getInputFirstName(), { target: { value: "Vadik" } });
      fireEvent.change(getInputLastName(), {
        target: { value: "Samitashvili" },
      });
      fireEvent.change(getInputEmail(), {
        target: { value: "totalgol2014@gmail.com" },
      });

      fireEvent.click(getButtonSave());

      await waitFor(() =>
        expect(onSubmit).toHaveBeenCalledWith({
          firstName: "Vadik",
          lastName: "Samitashvili",
          email: "totalgol2014@gmail.com",
        })
      );

      expect(onSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
