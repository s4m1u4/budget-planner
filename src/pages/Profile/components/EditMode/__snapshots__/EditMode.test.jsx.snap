// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit mode component edit mode make snapshot 1`] = `
<body
  style=""
>
  <div>
    <form
      action="#"
      style="width: 100%; max-width: 300px;"
    >
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="firstName"
        >
          First name
        </label>
        <input
          class="css-1pabijg"
          id="firstName"
          name="firstName"
          type="text"
          value="Vadym"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="css-1pabijg"
          id="lastName"
          name="lastName"
          type="text"
          value="Samitashvili"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="email"
        >
          Email
        </label>
        <input
          class="css-1pabijg"
          id="email"
          name="email"
          type="text"
          value="totalgol2015@gmail.com"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <button
        class="css-1pohnbb"
        color="secondary"
        style="margin-bottom: 1rem;"
        type="button"
      >
        Change password
      </button>
      <div
        class="css-1ijwswz"
      >
        <button
          class="css-1lrm4sh"
          color="success"
          disabled=""
          type="submit"
        >
          Save
        </button>
        <button
          class="css-3mo4ai"
          color="error"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
  <div>
    <form
      action="#"
      style="width: 100%; max-width: 300px;"
    >
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="firstName"
        >
          First name
        </label>
        <input
          class="css-1pabijg"
          id="firstName"
          name="firstName"
          type="text"
          value="Vadym"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="css-1pabijg"
          id="lastName"
          name="lastName"
          type="text"
          value="Samitashvili"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="email"
        >
          Email
        </label>
        <input
          class="css-1pabijg"
          id="email"
          name="email"
          type="text"
          value="totalgol2015@gmail.com"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <button
        class="css-1pohnbb"
        color="secondary"
        style="margin-bottom: 1rem;"
        type="button"
      >
        Change password
      </button>
      <div
        class="css-1ijwswz"
      >
        <button
          class="css-1lrm4sh"
          color="success"
          disabled=""
          type="submit"
        >
          Save
        </button>
        <button
          class="css-3mo4ai"
          color="error"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</body>
`;

exports[`Edit mode component regular mode make snapshot 1`] = `
<body>
  <div>
    <form
      action="#"
      style="width: 100%; max-width: 300px;"
    >
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="firstName"
        >
          First name
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="firstName"
          name="firstName"
          type="text"
          value="Vadym"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="lastName"
          name="lastName"
          type="text"
          value="Samitashvili"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="email"
        >
          Email
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="email"
          name="email"
          type="text"
          value="totalgol2015@gmail.com"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <button
        class="css-1pohnbb"
        color="secondary"
        style="margin-bottom: 1rem;"
        type="button"
      >
        Change password
      </button>
      <div
        class="css-1ijwswz"
      >
        <button
          class="css-2thluw"
          color="primary"
          type="button"
        >
          Edit
        </button>
      </div>
    </form>
  </div>
  <div>
    <form
      action="#"
      style="width: 100%; max-width: 300px;"
    >
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="firstName"
        >
          First name
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="firstName"
          name="firstName"
          type="text"
          value="Vadym"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="lastName"
        >
          Last name
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="lastName"
          name="lastName"
          type="text"
          value="Samitashvili"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <div
        class="css-acwcvw"
      >
        <label
          class="css-zkehil"
          for="email"
        >
          Email
        </label>
        <input
          class="css-1pabijg"
          disabled=""
          id="email"
          name="email"
          type="text"
          value="totalgol2015@gmail.com"
        />
        <p
          class="css-1ghz2lr"
        />
      </div>
      <button
        class="css-1pohnbb"
        color="secondary"
        style="margin-bottom: 1rem;"
        type="button"
      >
        Change password
      </button>
      <div
        class="css-1ijwswz"
      >
        <button
          class="css-2thluw"
          color="primary"
          type="button"
        >
          Edit
        </button>
      </div>
    </form>
  </div>
</body>
`;
